dallinger/config.py

```diff
@@ -0,0 +1,163 @@
+        self.extend(data, cast_types=True)
+
+    def load_from_environment(self):
+        self.extend(os.environ, cast_types=True)
+
+    def load_config(self):
+        if self.ready:
+            raise ValueError("Already loaded")
+        globalConfigName = ".dallingerconfig"
+        localConfigName = "config.txt"
+        if 'OPENSHIFT_SECRET_TOKEN' in os.environ:
+            globalConfig = os.path.join(os.environ['OPENSHIFT_DATA_DIR'], globalConfigName)
+        else:
+            globalConfig = os.path.expanduser(os.path.join("~/", globalConfigName))
+        localConfig = os.path.join(os.getcwd(), localConfigName)
+
+        defaults_folder = os.path.join(os.path.dirname(__file__), "default_configs")
+        local_defaults_file = os.path.join(defaults_folder, "local_config_defaults.txt")
+        global_defaults_file = os.path.join(defaults_folder, "global_config_defaults.txt")
+        if not os.path.exists(localConfig):
+            raise ValueError("No config.txt in the current directory")
+
+        # read default global and local, then user's global and local. This way
+        # any field not in the user's files will be set to the default value.
+        for config_file in [global_defaults_file, local_defaults_file, globalConfig, localConfig]:
+            self.load_from_config_file(config_file)
+        self.ready = True
+
+
+configurations = threading.local()
+
+
+def get_config():
+    if hasattr(configurations, 'config'):
+        return configurations.config
+    configurations.config = Configuration()
+
+    for key, type_, synonyms in (
+        ('ad_group', unicode, []),
+        ('amt_keywords', unicode, []),
+        ('anonymize_data', bool, []),
+        ('approve_requirement', int, []),
+        ('auto_recruit', bool, []),
+        ('aws_access_key_id', unicode, []),
+        ('aws_region', unicode, []),
+        ('aws_secret_access_key', unicode, []),
+        ('base_payment', float, []),
+        ('browser_exclude_rule', unicode, []),
+        ('clock_on', bool, []),
+        ('contact_email_on_error', unicode, []),
+        ('database_size', unicode, []),
+        ('database_url', unicode, []),
+        ('description', unicode, []),
+        ('duration', float, []),
+        ('dyno_type', unicode, []),
+        ('host', unicode, ['HOST']),
+        ('port', int, ['PORT']),
+        ('launch_in_sandbox_mode', bool, []),
+        ('lifetime', int, []),
+        ('logfile', unicode, []),
+        ('loglevel', int, []),
+        ('mode', unicode, []),
+        ('notification_url', unicode, []),
+        ('num_dynos_web', int, []),
+        ('num_dynos_worker', int, []),
+        ('num_participants', int, []),
+        ('organization_name', unicode, []),
+        ('psiturk_keywords', unicode, []),
+        ('table_name', unicode, []),
+        ('threads', unicode, []),
+        ('title', unicode, []),
+        ('us_only', bool, [])
+    ):
+        configurations.config.register(key, type_, synonyms)
```

Comment:
I can see the appeal of that, for sure. That would we couldn't have local experiment config have defaults in the global config, though, as it would raise an unknown key error. I guess that would happen anyway if you tried to run someone else's experiments.

@suchow Can I ask you about a hypothetical use-case? Might some people want to define new parameters in their experiments and then define values for them in their global config for local testing? I would guess this might be if one researcher had a lot of related experiments, or similar?

@davisagli Are you proposing the following flow?

1) Set up dallinger config variables
2) Scan config files, environment variables, etc for values and load them into the config
3) Allow the config to be read
4) Add experiment specific config variables
5) Rescan config files and environment variables
6) Allow local config to be read

If so, I am not sure I agree that the convenience of the dallinger entry points not having to explicitly saying the config should be loaded is balanced out by the convenience of not having to cause the configuration to be re-scanned after adding new variables.